### PR TITLE
21 créer une classe triplet image t image t+1 masque de différence

### PR DESCRIPTION
**Pull request by clement2323**
Initialement je voulais travailler sur les clases de type triple (t,t+1,label). J'ai créé ces classes puis lancé des entraînement sur le dataset S2Looking

Réorganisation des codes notamment création des sous dossiers datas et models https://github.com/ashleve/lightning-hydra-template/tree/main/src

Volonté de généraliser au mieux les datamodule/pl modules   (au max une classe par type de tâche segmentation detection)
Proposition de routine d'entraînement dans le train_segmentation.py (à discuter par zoom notamment sur la config et les limites de la généralisation), idée générale :

1) Les datasets prennent en entrée une liste de chemin pour les images et les labels et retournent systématiquement images, labels, dic avec (les images étant transformés en tensor puis augmentée, dioc contenant les chemins de fichiers pointant vers les originaux pour les futures représentations).

2) Les chargements des données et le découpage + controle des images/patchs sont fait en amont de l'entrainement (le temps est raisonnable par rapport à celui de l'entraînement).

Reste un effort de généralisation supplémentaire à fournir (à discuter par zoom très prochainement)

-> Ne pas prendre en compte les .ipynb (réorganisation légère des sous dossiers dans notebooks/)

Je me suis assuré que les codes était bien fomatés (merci flake 8)
J'ai géré les éventuels conflits avec un merge de la branche main au préalable avant la pull request

### CODE REVIEW

File: src/models/components/segmentation_models.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import torchvision
+from torch import nn
+
+
+class DeepLabv3Module(nn.Module):
+    """"""
+
```
Comment (tomseimandi) [inline src/models/components/segmentation_models.py:7]: J'y pense là mais pour les modèles aussi peut-être qu'on pourrait mettre une réf (papier ou autre) dans le docstring

File: src/utils/model_evaluation.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import torch
+
+# Calculate IOU
+
+
+def calculate_IOU(output, labels):
```
Comment (tomseimandi) [inline src/utils/model_evaluation.py:6]: Il faudrait ajouter un docstring pour documenter

**Review by tomseimandi — COMMENTED** [review]
Pas grand chose à dire comme tu nous avais montré comment ça marchait, il faut juste peut-être documenter un peu plus je pense

File: src/datas/components/change_detection_dataset.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+import numpy as np
+import torch
+from albumentations import Compose
+from torch.utils.data import Dataset
+
+from utils.satellite_image import SatelliteImage
+
+
+class end_to_end_cd_dataset(Dataset):
```
[the remark this answers is not in the document]
Comment (clement2323) [reply src/datas/components/change_detection_dataset.py:12]: bien noté, d'ailleurs je l'ai renommé avec le bon nom d'article -> ChangeIsEverywhereDataset

File: src/datas/components/change_detection_dataset.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        """
+        Constructor.
+
+        Args:
+            list_paths_images (List): list of path of the images
+            list_paths_labels (List): list of paths
+             containing the list_paths_labelss
+            transforms (Compose) : list of transforms
+        """
+        self.list_paths_images = list_paths_images
```
[the remark this answers is not in the document]
Comment (clement2323) [reply src/datas/components/change_detection_dataset.py:32]: Si, ils sont utilisés dans les lignes suivantes pour générer les listes d'image 1 et image 2 (resp. label1,2). 
En gros on a une liste de départ, on la mélange  -> list_image 1
On fait une rotation d'un pas en avant sur list_image_1 -> list_image_2. ce qui nous donne bien deux listes sur les quelles on peut appliquer un modèle de détection de changement, listes pour lesquelles on est sûr par construction qu'aucun élément ne se retrouve en face de lui même.

File: src/models/segmentation_module.py [hunk 5]
```diff
@@ -57,12 +45,12 @@ def __init__(
 
         self.model = model
         self.loss = nn.CrossEntropyLoss()
```
[the remark this answers is not in the document]
Comment (clement2323) [reply src/models/segmentation_module.py:47]: Oui je suis d'accord aussi je pense qu'il nous faudra un .py avec différents types de loss et un appele ne config

File: src/datas/components/change_detection_dataset.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        label = label.type(torch.LongTensor)
+
+        dic = {
+            "pathimage1": pathim1,
+            "pathimage2": pathim2,
+            "pathlabel1": pathlabel1,
+            "pathlabel2": pathlabel2,
+        }
+
+        return img_double, label, dic
```
[the remark this answers is not in the document]
Comment (clement2323) [reply src/datas/components/change_detection_dataset.py:134]: ça me va très bien, je m'en occupe !

File: src/models/components/segmentation_models.py (hunk 1 above)
Comment (clement2323) [reply src/models/components/segmentation_models.py:7]: done !

File: src/datas/__init__.py
[the remark this answers is not in the document]
Comment (tomseimandi) [reply src/datas/__init__.py:1]: Je pense que pour ça il faut juste remplacer la ligne dans le `.gitignore` par `/data/` !

File: src/data/components/change_detection_dataset.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+            sample2 = self.transforms(image=img2, label=label2)
+            img2 = sample2["image"]
+            label2 = sample2["label"]
+
+        else:
+            img1 = torch.tensor(img1.astype(float))
+            img1 = img1.permute([2, 0, 1])
+
+            img2 = torch.tensor(img2.astype(float))
+            img2 = img2.permute([2, 0, 1])
```
Comment (ThomasFaria) [inline src/data/components/change_detection_dataset.py:118]: Ici on aurait pas moyen de simplifier avec une petite fonction pour éviter tous les copier/coller avec im1 et im2 ? J'ai l'impression qu'on pourrait faire la même chose avec une fonction qui prend une image et un label en argument

File: src/train_detection_end_to_end.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+        "n channels train": 6,
+    }
+
+    config_train = {
+        "lr": 0.0001,
+        "momentum": 0.9,
+        "module": "deeplabv3",
+        "batch size": 2,
+        "max epochs": 100,
+    }
```
Comment (ThomasFaria) [inline src/train_detection_end_to_end.py:56]: C'est pas forcément pour de suite mais j'ai l'impression qu'on aura intérêt à mettre un kwargs ou autre dans les arguments de la fonction à terme. J'ai pas encore vu comment marche Argo workflows mais si on veut plus tard lancer plusieurs entraînements simultanément j'ai l'impression que ca sera plus simple avec des arguments plutôt que hard coded comme ca ou dans un fichier de config. @tomseimandi un avis ?

Comment (ThomasFaria) [reply src/train_detection_end_to_end.py:56]: Peut-être aussi enlever les espaces dans les keys des dico

File: src/train_detection_end_to_end.py [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+# python train_segementation
+# https://projet-slums-detection-561009.user.lab.sspcloud.fr
+#  segmentation testRaya
+
+remote_server_uri = (
+    "https://projet-slums-detection-561009.user.lab.sspcloud.fr"
+)
+experiment_name = "segmentation"
+run_name = "testraya_1"
+main(remote_server_uri, experiment_name, run_name)
```
Comment (ThomasFaria) [inline src/train_detection_end_to_end.py:230]: On enlève pour lancer les entrainements direct par bash ?
J'update le mlflow_train_segmentation.sh

File: src/train_segmentation.py
Comment (ThomasFaria) [inline src/train_segmentation.py:1]: Pas sur de voir la différence avec le fichier `train_detection_end_to_end.py`

**Review by ThomasFaria — APPROVED** [review]
Cool ça semble bien clean ! En grande partie ok pour moi juste quelques petites interrogations auxquelles vous pouvez répondre si vous n'en avez pas trop marre des reviews 😅

File: bash/mlflow_train_segmentation.sh [hunk 10]
```diff
@@ -17,6 +22,8 @@ while IFS= read -r line; do
 done <<< "$GET_PODS"
 
 export MLFLOW_TRACKING_URI="https://projet-slums-detection-$POD_ID.user.lab.sspcloud.fr"
-export MLFLOW_EXPERIMENT_NAME="segmentation"
 
-mlflow run ~/work/detection-bidonvilles/ --entry-point segmentation --env-manager=local -P remote_server_uri=$MLFLOW_TRACKING_URI
+mlflow run ~/work/detection-bidonvilles/ --entry-point segmentation --env-manager=local \
```
Comment (ThomasFaria) [inline bash/mlflow_train_segmentation.sh:26]: @tomseimandi si on définit un entry point ici, ca veut dire que dans notre fichier train_... .py au lien de :
`if __name__ == "__main__":` il faut mettre `if __name__ == "segmentation":` ou rien à voir ?

Comment (tomseimandi) [reply bash/mlflow_train_segmentation.sh:26]: ça correspond aux entry-points du fichier MLproject ça

File: src/train_detection_end_to_end.py (hunk 8 above)
Comment (tomseimandi) [reply src/train_detection_end_to_end.py:56]: Oui je suis d'accord !
